keps/sig-node/375-cpu-manager/README.md

```diff
@@ -323,6 +395,240 @@ func (p *staticPolicy) RemoveContainer(s State, containerID string) error {
+
+#### Deprecation
+
+- Announce deprecation and support policy of the existing flag
+- Two versions passed since introducing the functionality that deprecates the flag (to address version skew)
+- Address feedback on usage/changed behavior, provided on GitHub issues
+- Deprecate the flag
+-->
+
+### Upgrade / Downgrade Strategy
+
+No impact. It's always possible to trivially downgrade to the previous kubelet
+
+### Version Skew Strategy
+
+Not relevant
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [ ] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name:
+  - Components depending on the feature gate:
+- [ ] Other
+  - Describe the mechanism:
+  - Will enabling / disabling the feature require downtime of the control
+    plane?
+  - Will enabling / disabling the feature require downtime or reprovisioning
+    of a node? (Do not assume `Dynamic Kubelet Config` feature is enabled).
+
+###### Does enabling the feature change any default behavior?
+
+No, unless the non-none policy is explicitly configured.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes, using the kubelet config.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The impact is node-local only.
+If the state of a node is steady, no changes.
+If a guaranteed pod is admitted, running non-guaranteed pods will have their CPU cgroup changed while running.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes, covered by e2e tests
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+A rollout can fail if a bug in the cpumanager prevents _new_ pods to start, or existing pods to be restarted.
+Already running workload will not be affected if the node state is steady
+
+###### What specific metrics should inform a rollback?
+
+Pod creation errors o a node-by-node basis.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+No to both.
+Changes in behavior only affects pods meeting the conditions (guaranteed QoS, integral CPU request) scheduler after the upgrade.
+Running pods will be unaffected by any change. This offers some degree of safety in both upgrade->rollback
+and upgrade->downgrade->upgrade scenarios.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+Monitor the pod admission counter
+Monitor the pods not going running after successful schedule
+
+###### How can an operator determine if the feature is in use by workloads?
+
+The operator need to inspect the node and verify the cpu pinning assignment either checking the cgroups on the node
```

Comment:
CPU pinning is indicated on the pod somewhere, right?  Which field on the pod indicates usage?

Reply:
the request of cpu pinning is requested by the pod

1. opting in the "guaranteed" QoS
2. requesting integral number of CPUs

there is no single value (e.g. field in pod status) which however report if pinning was succesfull. Working on it. *Can* be a gap, will check and report.

Reply:
@deads2k cpu pinning is a node local decision.

if the node is configured to support the feature (which is not advertised to the scheduler because its a node local optimization), and the pod is in the guaranteed qos class and advertises integral cpus, the kubelet if configured will pin the pod to dedicated cpus.

there is no 'pin this pods cpu' feature advertised in the pod API itself by design.

Reply:
The resources request: `pod.spec.containers[*].resources.limits.cpu`. If it's integral value (e.g. 2) and if the pod QoS is guaranteed, this combination will trigger the exclusive CPU allocation and the pinning (https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy)